Definitions.md

```diff
@@ -111,28 +111,79 @@ Runtime subsystem(s) and processes that handle the various operations for import
 
 > **Note**: This is ideally where realm-based module remapping would take place.
 
+```js
+/**
+ * @param toImport exact string argument to `import`
+ * @param referrer base moduleId for the importing module
+ * @return moduleId
+ */
+function resolve(toImport: string, referrer: string): string;
+```
+
 <dt>Location
 <dd>The operation used to determine the unique resource location of a module based on its unique module identifier, by deterministic derivation, mapping and/or device-dependent operations, applicable to non-synthetic modules and synthetically wrapped resource modules which are retrieved from externally.
 
-> **Note**: This is where context-based pre-caching would take place.
+> **Note**: This is where context-based cache warming would take place.
+
+```js
+/**
+ * @param moduleId the unique module identifier to locate
+ * @return URL
+ */
+function locate(moduleId: string): string
+```
 
 <dt>Retrieval
 <dd>The operation used to transport content of non-synthetic and synthetically wrapped resource modules, usually handled on a per-context basis.
 
 > **Note**: This is where context-based security/consistency checking would take place, and would likely include source text parsing and static pre-linking (ie for all relatively scoped resources).
 
-<dt>Linking
+```js
+// This record type is implementation-defined.
+type ModuleLinkageRecord = { moduleFunctor: string };
 
-<dd>The operation used to normalize and/or synthesize the single context or realm specific manifestation of a module, including the namespace object which hold the prefined set of yet-to-be-initialized exports.
+/**
+ * @param url the URL to retrieve
+ * @return Record describing module's linking requirements
+ */
+function retrieve(url: string): Promise<ModuleLinkageRecord>
+```
 
+<dt>Linking
+
+<dd>The operation used to normalize and/or synthesize the single context or realm specific manifestation of a module, including the namespace object which hold the predefined set of yet-to-be-initialized exports.
+  
 > **Note**: This is ideally where augmentation (ie instrumentation) would take place.
 
+```js
+// This record type can contain additional information.
+type ModuleInstance = {
+  initialize(): Promise<ModuleNamespace>;
+};
+
+/**
+ * @param moduleLinkageMap map from a moduleId to its ModuleLinkageRecord
+ * @param moduleId the root of this linkage graph
+ * @param evaluator the evaluation function for the moduleFunctor
+ * @param preEndowments module-specific endowments for the moduleFunctor
```

Comment:
`preEndowments` should be `moduleEndowmentsMap` which is a map from the `moduleId` to its endowment overrides.